File: 프로그래머스/2/92341. 주차 요금 계산/주차 요금 계산.py
def solution(fees, records):
    answer = []
    record_hash = {}
    fee_hash = {}   # {차량번호: [주차시간, 요금]}
    
    # 1. 총 주차 시간 계산
        # record_hash = {차량번호 : [시간, 플래그]} 만들기.
    for r in records:
        time, num, flag = r.split()
        h, m = time.split(':')
        h, m = int(h), int(m)
        time = h*60 + m     # 입차시간
        
        # 입차: 입차시간 record_hash 저장 -> fee_hash 초기화
        if num not in record_hash:
            record_hash[num] = [time, flag]
            fee_hash[num] = [0, fees[1]]
        # 출차: 주차 시간 계산 -> fee_hash 저장 -> record_hash flag 변경
        elif flag == "OUT":    
            park_time = time - record_hash[num][0]
            fee_hash[num][0] += park_time
            record_hash[num][1] = flag
        # 재입차: 입차 시간 record_hash 저장 -> record_hash flag 변경
        elif flag == "IN":      # 입차
            record_hash[num][0] = time
            record_hash[num][1] = flag

    # 2. 입차 후 출차 안한 차 확인
    for num in record_hash.keys():
        if record_hash[num][1] == "IN":
            fee_hash[num][0] += 1439 - record_hash[num][0]
    
    # 3. 가격 계산
        # fee_hash 순회 -> ((총 주차시간-기본시간-1) // 단위 시간 + 1) * 단위 요금

    for num in fee_hash.keys():
        park_time = fee_hash[num][0] - fees[0]
        if park_time <= 0:
            continue
        else:
            t = ((park_time - 1) // fees[2] + 1)
            fee_hash[num][1] +=  t * fees[3]
    
    fee_hash = sorted(fee_hash.items(), key=lambda x: x[0])
    answer = [car[1][1] for car in fee_hash]
    return answer

File: 프로그래머스/2/92341. 주차 요금 계산/주차 요금 계산.swift
import Foundation
/**
- 입차 후 출차 내역 없다면 23:59에 출차한 것.
- 초과한 시간이 단위 시간으로 나누어 떨어지지 않으면 올림
- 차량 번호가 작은 자동차부터 주차 요금 리턴
- 요금은 하루 주차 시간을 통해 "일괄 정산"

로직
1. 차 번호 별 총 주차 시간 구하기 => [차량 번호: 총 주차 시간] 딕셔너리에 저장
    1) 해당 차량의 OUT이 나올 때까지 IN 기록 저장 필요 => [차량 번호: 입차 시간] 딕셔너리에 저장
    2) OUT이 나온다면 [차량 번호] = nil 저장 -> 시간 계산해서 [차량 번호: 총 주차 시간] 딕셔너리 갱신
2. 총 주차 시간 -> 요금 변환 => [차량 번호: 요금] 딕셔너리에 저장
3. 결과 배열 리턴
*/



/// fees: [기본 시간, 기본 요금, 단위 시간, 단위 요금], records: [시각, 차량번호, 내역]
func solution(_ fees:[Int], _ records:[String]) -> [Int] {
    var totalFee = [String: Int]()
    var result = [Int]()
    
    // 총 주차 시간 구하기
    let totalParkingTime = getTotalParkingTime(records)
    print(totalParkingTime)
    // 요금 변환
    for (carNum, time) in totalParkingTime {
        totalFee[carNum] = calculateFee(time, fees)
    }
    
    for carNum in totalFee.keys.sorted() {
        let fee = totalFee[carNum, default: 0]
        result.append(fee)
    }

    return result
}

func parseRecord(_ record: String) -> (Int, String, Bool) {
    let components = record.components(separatedBy: " ")
    let time = parseTime(components[0])
    let carNum = components[1]
    let isIn = components[2] == "IN" ? true : false
    return (time, carNum, isIn)
}

func parseTime(_ time: String) -> Int {
    let time = time.components(separatedBy: ":")
    let (h, m) = (Int(time[0])!, Int(time[1])!)
    return h * 60 + m
}

func getTotalParkingTime(_ records: [String]) -> [String: Int] {
    var enterTime = [String: Int]()            // [차량 번호: 입차 시간]
    var totalParkingTime = [String: Int]()     // [차량 번호: 총 주차 시간]

    for record in records {
        let (time, carNum, isIn) = parseRecord(record)
        
        if isIn {
            enterTime[carNum] = time
        } else {
            let parkingTime = time - enterTime[carNum, default: 0]
            totalParkingTime[carNum, default: 0] += parkingTime
            enterTime[carNum] = nil
        }
    }

    for (carNum, time) in enterTime {
        totalParkingTime[carNum, default: 0] += parseTime("23:59") - time
    }

    return totalParkingTime
}

/// fees: [기본 시간, 기본 요금, 단위 시간, 단위 요금]
func calculateFee(_ time: Int, _ fees: [Int]) -> Int {
    if time <= fees[0] {
        return fees[1]
    } else {
        let additionalFee = (((time - fees[0] - 1) / fees[2]) + 1) * fees[3]
        return fees[1] + additionalFee
    }
}